packages/tenant-process/src/services/tenantService.ts

```diff
@@ -1744,141 +1745,212 @@ export function tenantServiceBuilder(
         readModelService
       );
 
-      assertFeatureNotAssigned(requesterTenant.data, "DelegatedProducer");
+      const delegatedConsumerEvent = match(
+        tenantFeatures.isDelegatedConsumerFeatureEnabled
+      )
+        .with(true, () =>
+          assignTenantDelegatedConsumerFeature({
+            tenant: requesterTenant,
+            correlationId,
+            logger,
+          })
+        )
+        .with(false, () =>
+          removeTenantDelegatedConsumerFeature({
+            tenant: requesterTenant,
+            correlationId,
+            logger,
+          })
+        )
+        .exhaustive();
 
-      const updatedTenant: Tenant = {
-        ...requesterTenant.data,
-        features: [
-          ...requesterTenant.data.features,
-          { type: "DelegatedProducer", availabilityTimestamp: new Date() },
-        ],
-        updatedAt: new Date(),
-      };
+      const updatedTenant: WithMetadata<Tenant> = delegatedConsumerEvent
+        ? {
+            ...requesterTenant,
+            data: {
+              ...delegatedConsumerEvent.updatedTenant,
+            },
+          }
+        : requesterTenant;
 
-      await repository.createEvent(
-        toCreateEventTenantDelegatedProducerFeatureAdded(
-          requesterTenant.metadata.version,
-          updatedTenant,
-          correlationId
+      const delegatedProducerEvent = match(
+        tenantFeatures.isDelegatedProducerFeatureEnabled
+      )
+        .with(true, () =>
+          assignTenantDelegatedProducerFeature({
+            tenant: updatedTenant,
+            correlationId,
+            logger,
+          })
         )
-      );
+        .with(false, () =>
+          removeTenantDelegatedProducerFeature({
+            tenant: updatedTenant,
+            correlationId,
+            logger,
+          })
+        )
+        .exhaustive();
+
+      await match([delegatedConsumerEvent, delegatedProducerEvent])
+        .with(
+          [P.not(P.nullish), P.not(P.nullish)],
+          async ([delegatedConsumerEvent, delegatedProducerEvent]) =>
+            repository.createEvents([
+              delegatedConsumerEvent.event,
+              {
+                ...delegatedProducerEvent.event,
+                version: requesterTenant.metadata.version + 1,
+              },
+            ])
+        )
+        .with([P.not(P.nullish), null], async ([delegatedConsumerEvent, _]) =>
+          repository.createEvent(delegatedConsumerEvent.event)
+        )
+        .with([null, P.not(P.nullish)], async ([_, delegatedProducerEvent]) =>
+          repository.createEvent(delegatedProducerEvent.event)
+        )
+        .run();
```

Comment:
Let's always use `.exhaustive()` with TS pattern, so that we avoid runtime failures due to potential non-exhaustive matches

```suggestion
        .exhaustive();
```